Unify passing sampling rate information

cczhu opened the pull request:
Replaced `frames_per_second` with a `sample_rate` property from all stream reader and writer classes.  Correspondingly altered time and rate calculations, interfaces to headers with sampling and framerates, and tests, docstrings and documentation examples.

(Also removed redundant size definition in Mark 5B.)

Addresses #109.

mhvk commented:
Fine to remove VDIF header bandwidth and framerate, replace by samplerate.

cczhu commented:
I just remembered - currently `__repr__` prints out `_sample_rate` as a float, with "Hz" tacked on afterward.    I'm not sure if `astropy.units` quantities naturally have a way to print a set number of significant figures and append a unit.  Ideally we'd switch between Hz and MHz whenever it would be more compact to use the latter, but I don't know if there's functionality for this already in `astropy.units`.

mhvk commented:
I've added a comment in-line -- this will become visible when I submit the review.

mhvk commented on baseband/gsb/base.py:
```diff
@@ -167,7 +179,7 @@ def __repr__(self):
         return ("<{s.__class__.__name__} header={s.fh_ts.name}"
                 " offset= {s.offset}\n    data={dn}\n"
-                "    frames_per_second={s.frames_per_second:.3f},"
+                "    sample_rate={s._sample_rate:.3e} Hz,"
```
p.s. No way yet to auto-convert to the "nicest" unit, but a PR is in the works: https://github.com/astropy/astropy/pull/7033

mhvk commented on baseband/gsb/base.py:
```diff
@@ -142,12 +141,25 @@ def __init__(self, fh_ts, fh_raw, header0, thread_ids=None,
+                header1 = self.fh_ts.read_timestamp()
+                assert (fh_ts.tell() ==
+                        header0.seek_offset(2))
+                sample_rate = samples_per_frame * (
```
I wonder if we should use knowledge of the GSB here: the "master" clock is at 200 MHz, and sample rates are integer divisions of this... (fine for another PR)

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -436,6 +441,12 @@ def test_raw_stream(self, tmpdir):
             assert fh_r.size == 10 * fh_r.samples_per_frame
+            # Sample rates aren't exactly equal because GSB timestamps only
+            # store 9 digits of the fractional second.
+            assert np.all(np.isclose(fh_r.sample_rate, sample_rate,
```
See comment above; thinking further about this: this will screw up actual times and thus make VLBI nearly impossible -- we do need to be sure this is right...  I wonder if we should remove auto-guessing for now and do that correctly in another PR.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -436,6 +441,12 @@ def test_raw_stream(self, tmpdir):
+            # store 9 digits of the fractional second.
+            assert np.all(np.isclose(fh_r.sample_rate, sample_rate,
+                                     rtol=1e-9, atol=0.*u.Hz))
+            assert np.isclose(fh_r._sample_rate, sample_rate.value,
```
No testing of private properties unless they are really different from the public ones.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -527,7 +541,7 @@ def test_raw_stream(self, tmpdir):
         with gsb.open(filename_incompletehead, mode='rs', raw=SAMPLE_RAWDUMP,
                       payloadsize=2**12, squeeze=False,
-                      frames_per_second=3.9736) as fh_r:
+                      sample_rate=(self.framerate * 2**13 * u.Hz)) as fh_r:
```
`sample_rate` is defined here, I think? Maybe have `self.sample_rate`?

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -548,6 +563,12 @@ def test_phased_stream(self, tmpdir):
             assert fh_r.header0.time == fh_r.start_time
             assert fh_r.header0 == frame1.header
             assert fh_r.size == 10 * fh_r.samples_per_frame
+            # Sample rates aren't exactly equal because GSB timestamps only
```
As above

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -605,22 +626,26 @@ def test_phased_stream(self, tmpdir):
-                            frames_per_second=fh_r.frames_per_second,
+                            sample_rate=(self.framerate * 2**3 * u.Hz),
                             **fh_r.header0)
+            assert np.all(np.isclose(fh_w.sample_rate, sample_rate,
```
Since we're not guessing here, shouldn't this just be equal?

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -475,20 +486,23 @@ def test_raw_stream(self, tmpdir):
                 assert fh_n.size == fh_r.size
                 assert fh_n.sample_shape == fh_r.sample_shape
                 assert fh_n.start_time == fh_r.start_time
+                assert np.all(np.isclose(fh_n.sample_rate, sample_rate,
```
Since we're not guessing, shouldn't this just be equal?

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -605,22 +626,26 @@ def test_phased_stream(self, tmpdir):
                 assert fh_n.size == fh_r.size
                 assert fh_n.sample_shape == fh_r.sample_shape
                 assert fh_n.start_time == fh_r.start_time
+                assert np.all(np.isclose(fh_n.sample_rate, sample_rate,
```
Same, not guessing here.

mhvk commented on baseband/mark4/base.py:
```diff
@@ -304,32 +303,33 @@ def __init__(self, fh_raw, ntrack=None, decade=None, ref_time=None,
             Binary file handle.
         header_template : header class or instance
             Definition or instance of file format's header class.
+        samples_per_frame : int
```
This is guaranteed to be available from the header, so no need to pass it in.

mhvk commented on baseband/mark4/base.py:
```diff
@@ -341,7 +341,7 @@ def _get_frame_rate(fh, header_template):
         # to 160 ms.
         tdelta = header1.ms[0] - header0.ms[0]
-        return int(np.round(1000. / tdelta))
+        return int(np.round(1000. / tdelta)) * samples_per_frame * u.Hz
```
Just replace with `header0.samples_per_frame`

mhvk commented on baseband/mark4/tests/test_mark4.py:
```diff
@@ -505,6 +505,8 @@ def test_filestreamer(self, tmpdir):
             assert fh.samples_per_frame == 80000
             assert fh.size == 2 * fh.samples_per_frame
+            assert np.all(fh.sample_rate == 32 * u.MHz)
+            assert fh._sample_rate == 32000000.0
```
no private property tests and no `np.all`, like for GSB

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -348,17 +336,16 @@ class Mark5BStreamWriter(VLBIStreamWriterBase, Mark5BFileWriter):
 
-    def __init__(self, raw, frames_per_second=None, sample_rate=None,
-                 nchan=1, bps=2, header=None, squeeze=True, **kwargs):
+    def __init__(self, raw, sample_rate, nchan=1, bps=2, header=None,
```
This makes me wonder whether `sample_rate` should be first generically...

cczhu replied:
From above, `sample_rate` is one of two optional variables in `VLBIStreamReaderBase.__init__`, which is why it's second-to-last.  Of course, we could make `sample_rate` a positional argument, but I like the fact that all the positional arguments are values `VLBIStreamReaderBase` doesn't know what to do with if they weren't explicitly passed.  If you feel having `sample_rate` be the first keyword makes the code more readable, we can change it (though I'd then argue `samples_per_frame`, whenever passed, should go right after).

mhvk commented on baseband/mark5b/tests/test_mark5b.py:
```diff
@@ -345,7 +345,8 @@ def test_filestreamer(self, tmpdir):
             assert fh.fh_raw.tell() == header.framesize
             assert fh.samples_per_frame == 5000
-            assert fh.frames_per_second == 6400
+            assert np.all(fh.sample_rate == 32 * u.MHz)
```
No `np.all`, no testing of private properties.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -238,14 +238,14 @@ class VDIFStreamBase(VLBIStreamBase):
+        if sample_rate is None:
             try:
-                frames_per_second = int(header0.framerate.to(u.Hz).value)
+                sample_rate = samples_per_frame * header0.framerate.to(u.Hz)
```
This will now become `header0.sample_rate()` -- much nicer!

mhvk commented on baseband/vdif/base.py:
```diff
@@ -255,20 +255,20 @@ def __init__(self, fh_raw, header0, thread_ids, frames_per_second=None,
         This passes on frame rate, since not all VDIF headers can calculate it.
         """
-        return header.get_time(framerate=self.frames_per_second * u.Hz)
+        return header.get_time(
```
Can the header work with `sample_rate`? Through maybe this is fine.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -255,20 +255,20 @@ def __init__(self, fh_raw, header0, thread_ids, frames_per_second=None,
     def __repr__(self):
         return ("<{s.__class__.__name__} name={s.name} offset={s.offset}\n"
-                "    frames_per_second={s.frames_per_second},"
+                "    sample_rate={s._sample_rate} Hz,"
```
As for GSB.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -444,41 +442,46 @@ class VDIFStreamWriter(VDIFStreamBase, VLBIStreamWriterBase, VDIFFileWriter):
-    bandwidth : `~astropy.units.Quantity`
-        In frequency units.  Sufficient for `edv` 1, 3, or 4 to determine the
-        frames per second.
+    framerate : `~astropy.units.Quantity`, optional
```
Will be nice to get these in terms of `sample_rate` as well.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -444,41 +442,46 @@ class VDIFStreamWriter(VDIFStreamBase, VLBIStreamWriterBase, VDIFFileWriter):
+            raw, header, range(nthread), sample_rate=sample_rate,
+            squeeze=squeeze)
         # Set framerate and thus bandwidth in the header, if not set already.
         try:
```
Another thing that will become nicer with `header.samplerate`

cczhu replied:
I appended an assert line to the try/except/else block to ensure that the header's `sample_rate` is same as the stream reader's (it's possible to pass a complete `header0`, or keywords `sampling_rate` and `sampling unit`, while separately passing a `sample_shape`).

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -496,13 +496,15 @@ def test_filestreamer(self):
             assert repr(fh).startswith('<VDIFStreamReader')
             assert fh.tell() == 0
             assert header == fh.header0
+            assert np.all(fh.sample_rate == 32 * u.MHz)
```
Just so it doesn't get missed: no `np.all`, no testing of private properties

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -599,6 +602,37 @@ def test_stream_writer(self, tmpdir):
             assert np.all(fh.read() == record)
 
+        # Test writing a file using header keywords (in particular
+        # ``bandwidth`` and ``framerate`` to set the sample rate).
```
So this may disappear again... sorry.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -49,27 +49,19 @@ class VLBIStreamBase(VLBIFileBase):
 
     _frame_class = None
     _squeezed_shape = None
+    _sample_rate = None
```
I don't think this is needed, in the end.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -167,6 +159,15 @@ def thread_ids(self):
     def thread_ids(self, ids):
         self._thread_ids = ids
 
+    @property
```
I think for now I would do without the getter and setter, and just make this an attribute... Eventually, for proper VLBI classes we probably want to check `sample_rate` is consistent with an integer frame rate.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -189,18 +190,18 @@ def tell(self, unit=None):
 
         return (self.offset * u_sample).to(unit, equivalencies=[(u.s, u.Unit(
-            self.samples_per_frame * self.frames_per_second * u_sample))])
+            self._sample_rate * u_sample))])
```
Use `self.sample_rate.to_value(u.Hz) * u_sample`.  Though I actually forget why one couldn't just do `(self.offset / self.sample_rate).to(unit)`. Can you try that?

cczhu replied:
The latter does work!

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -189,18 +190,18 @@ def tell(self, unit=None):
-        full_frame_nr, extra = divmod(offset, self.samples_per_frame)
-        dt, frame_nr = divmod(full_frame_nr, self.frames_per_second)
-        return int(dt), int(frame_nr), extra
+        dt, subsecond_samples = divmod(offset, self._sample_rate)
```
One place where `_sample_rate` is functional. I'd still just do `self.sample_rate.to_value(u.Hz)` (faster than `to(u.Hz).value`, but may need newer astropy...)

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -189,18 +190,18 @@ def tell(self, unit=None):
     def __repr__(self):
         return ("<{s.__class__.__name__} name={s.name} offset={s.offset}\n"
-                "    frames_per_second={s.frames_per_second},"
+                "    sample_rate={s._sample_rate} Hz,"
```
just `{s.sample_rate}` will work.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -211,54 +212,55 @@ def __repr__(self):
-                frames_per_second = self._get_frame_rate(fh_raw,
-                                                         header0)
+                sample_rate = self._get_sample_rate(fh_raw, header0,
+                                                    samples_per_frame)
```
Ah, now I see the possible pitfall: do all headers have `samples_per_frame` -- which is not true for Mark 5B.  How about just keeping this as `_get_frame_rate` -- after all, that is what the basic routine does if times fail it -- and then multiply with `samples_per_frame` here.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -301,7 +303,10 @@ def stop_time(self):
         the time of the sample pointer's current offset.
         """
-        return self._get_time(self._last_header) + u.s / self.frames_per_second
+        # Faster to use self._sample_rate and append units than to convert
```
Don't bother -- this is not a slow part of the process.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -314,7 +319,7 @@ def get_time1(self):
     def size(self):
         """Number of samples in the file."""
         return int(round((self.stop_time - self.start_time).to(u.s).value *
```
Use `sample_rate` so you can just do `int(round((self.stop_time - self.start_time) * self.sample_rate))`

cczhu replied:
I'm not sure if this is due to my environment (I'm on Py3.5.2, running numpy 1.14.0 and astropy 2.0.3), but this requires `np.round` and decomposing the units before casting to `int`, i.e.
```
        return int(np.round(((self.stop_time - self.start_time) *
                             self.sample_rate).decompose()))
```

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -345,7 +350,7 @@ def seek(self, offset, whence=0):
 
             offset = offset.to(u_sample, equivalencies=[(u.s, u.Unit(
-                self.samples_per_frame * self.frames_per_second * u_sample))])
+                self._sample_rate * u_sample))])
```
Here, like for `tell`, I think on second sight, one can simplify hugely and just do
```
offset = int(round(offset * self.sample_rate))
```

That probably means we do not need `u_sample` any more...

cczhu replied:
As above.

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -430,7 +431,7 @@ def test_streambase_squeeze(self):
                                  'n1, n2, n3, n4, n5, n6, n7, n8')
         sample_shape = smp_shp_cls(1, 17, 3, 2, 1, 5, 1, 1)
         sb = VLBIStreamBase(None, None, sample_shape, 1, False,
```
Looking at this test again, I think we should use keyword arguments -- I've got no idea what it is doing!!

mhvk commented on docs/baseband/dada/index.rst:
```diff
@@ -49,7 +50,7 @@ writing is in units of samples, and provides access to header information.
     >>> fh
     <DADAStreamReader name=... offset=0
-        frames_per_second=1000.0, samples_per_frame=16000,
+        sample_rate=16000000.0 Hz, samples_per_frame=16000,
```
I hope this will change to `16.0 MHz`...

mhvk commented on docs/baseband/dada/index.rst:
```diff
@@ -67,7 +68,6 @@ even smaller size of the payload, to show how one can define multiple files.
-    >>> import astropy.units as u
     >>> fw = dada.open('{utc_start}.{obs_offset:016d}.000000.dada', 'ws',
     ...                npol=2, samples_per_frame=5000, nchan=1, bps=8,
     ...                bandwidth=16*u.MHz, complex_data=True,
```
Maybe remove `bandwidth` in this example too...

mhvk replied:
Still outstanding.

mhvk commented on docs/baseband/tutorials/new_edv.rst:
```diff
@@ -430,7 +431,7 @@ class, and define a replacement::
 We can then use the stream reader without further modification::
 
-    >>> fh2 = vdif.open(SAMPLE_DRAO_CORRUPT, 'rs', frames_per_second=390625)
+    >>> fh2 = vdif.open(SAMPLE_DRAO_CORRUPT, 'rs', sample_rate=625**3*u.Hz)
```
This seems wrong: `625**2` or `5**6`?

cczhu commented on baseband/gsb/base.py:
```diff
@@ -111,8 +111,7 @@ class GSBStreamBase(VLBIStreamBase):
                  samples_per_frame=None, payloadsize=None,
-                 frames_per_second=None, sample_rate=None,
-                 squeeze=True):
+                 sample_rate=None, squeeze=True):
```
`VLBIStreamReaderBase.__init__` sets default values only for `sample_rate` and `squeeze`, hence why `sample_rate` is the second last argument for all readers.  I realized the order in the docstrings isn't consistent with the code, and have fixed this for all formats.

For GSB, however, we discussed yesterday removing the ability to auto-calculate the frame rate, since timestamps from the rawdump headers may drift by up to 10 ms (according to the GSB documentation on phased timestamps...).  I've thus made `sample_rate` a required argument, and adjusted the docstring as well.

mhvk replied:
I think having it at the end consistently is perhaps more important. Maybe after all use that GSB really can have only a few settings? Anyway, find to leave it near the end here.

cczhu commented on baseband/gsb/base.py:
```diff
@@ -142,12 +141,25 @@ def __init__(self, fh_ts, fh_raw, header0, thread_ids=None,
             sample_shape = GSBPayload._sample_shape_maker_nthread(len(fh_raw),
                                                                   nchan)
 
+        if sample_rate is None:
```
As per our discussion yesterday, I've removed the ability for GSB readers to auto-calculate sample_rate.

cczhu commented:
Per our quick discussion, I'm keeping `sample_rate` a property to be documented by Sphinx, but storing a quantity rather than a float.  I did the same to `samples_per_frame`.  I also revised `_frame_info` to first divide by `self.samples_per_frame` and then `framerate`.  I've removed GSB's ability to auto-calculate `frames_per_second`.  VDIF's `bandwidth` and `framerate` header properties have now been replaced by a `sample_rate` property.

cczhu commented:
Okay, so older versions of Astropy can't handle `to_value`.  I'll replace those with `to(u.Hz).value`.

cczhu commented:
I discovered that `to_value(u.Hz)` is not supported by Astropy 1.3.x, so replaced `to_value(u.Hz)` with `to(u.Hz).value`.

Strangely, Travis CL ran successfully back when I used `to_value`, so it must not be testing against Astropy 1.x.x.  We can either change our dependencies, or add older versions of Astropy to Travis.

mhvk commented:
No reason to support astropy 1.3 for the real baseband release. Astropy 2.0 is the current long-time support release (and last that supports python2). So, I think the travis setup we have is OK.

cczhu commented:
In that case I'll revert to `to_value` before we merge.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -167,6 +158,30 @@ def thread_ids(self):
+        """Number of complete samples per frame."""
+        return self._samples_per_frame
+
+    @samples_per_frame.setter
```
Since the setter does not do anything, I would for now remove it, and just set `self._samples_per_frame` in `__init__`. That means you also do not have to set it to `None` in the class.

Alternatively, use a try/except around `self._samples_per_frame = samples_per_frame.__index__()`, i.e., check that it is (convertable to) a proper integer.

cczhu replied:
I like the latter - implementing.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -188,19 +203,19 @@ def tell(self, unit=None):
     def _frame_info(self):
         offset = (self.offset +
                   self.header0['frame_nr'] * self.samples_per_frame)
+        framerate = self.sample_rate.to_value(u.Hz) / self.samples_per_frame
```
This is really VLBI only so we can assume the framerate itself is integer, i.e.,
```
frames_per_second = int(round((self.sample_rate / self.samples_per_frame).to(u.Hz)))
```
and then we no longer have to cast anything else to `int` below.

mhvk replied:
OK as per the comment above, but for each maybe just `to_value(u.Hz)` to avoid the issue with `round`

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -188,19 +203,19 @@ def tell(self, unit=None):
     def __repr__(self):
         return ("<{s.__class__.__name__} name={s.name} offset={s.offset}\n"
-                "    frames_per_second={s.frames_per_second},"
+                "    sample_rate={s.sample_rate:.6g},"
```
I would not do the `.6g` here since I think generally the numbers will be OK. Only really necessary for GSB.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -313,8 +328,8 @@ def get_time1(self):
         """Number of samples in the file."""
-        return int(round((self.stop_time - self.start_time).to(u.s).value *
-                         self.frames_per_second * self.samples_per_frame))
+        return int(np.round(((self.stop_time - self.start_time) *
```
Strange that one needs round... That's a buglet in Quantity. But don't use decompose, but rather the unit you want: Maybe `int(((self.stop_time - self.start_time) * self.sample_rate).to(u.one).round())`. (or `u.dimensionless_unscaled` if you want to be more elaborate...)

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -344,9 +359,10 @@ def seek(self, offset, whence=0):
+            # offset = offset.to(u_sample, equivalencies=[(u.s, u.Unit(
+            #     self._sample_rate * u_sample))])
+            # offset = int(round(offset.value))
+            offset = int(np.round((offset * self.sample_rate).decompose()))
```
Same here, `to(u.one)`. Also, no commented out code!!

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -355,7 +371,7 @@ def seek(self, offset, whence=0):
             self.offset = self.size + offset
         else:
-            raise ValueError("invalid 'whence'; should be 0 or 'start', 1 or"
+            raise ValueError("Invalid 'whence'; should be 0 or 'start', 1 or"
```
The standard in python actually seems to be not to capitalize the first letter of an exception string (because there is a colon in front, not a period). E.g.,
```
hash([1,2])
TypeError: unhashable type: 'list'
```

cczhu replied:
I see; this is apparently also the case for `warnings.warn`:
```
warnings.warn("Test warning")
/home/cczhu/pythonenv/ml3/bin/ipython3:1: UserWarning: Test warning
```
Though both capital and lower case are used throughout baseband.  I've made the change in the files I'm currently working on, but I'll also raise a tiny issue that this should be standardized for 1.0 (just so I don't have to search through every file for errors for this PR).

mhvk replied:
Yes, I only became aware of this relatively recently myself, and so the package is quite inconsistent. Obviously, not that big a deal!

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -440,7 +444,7 @@ def test_streambase_squeeze(self):
                                    'n1')
         sample_shape_short = smp_shp_cls_s(1)
         sbs = VLBIStreamBase(None, None, sample_shape_short, 1, False,
```
I'd prefer keywords here too...

mhvk commented on docs/baseband/mark4/index.rst:
```diff
@@ -48,7 +50,7 @@ also provide a reference time within 4 years of the observation start time::
     >>> fh
     <Mark4StreamReader name=... offset=0
-        frames_per_second=400, samples_per_frame=80000,
+        sample_rate=3.2e+07 Hz, samples_per_frame=80000,
```
guess you hadn't gotten to the doc yet...

cczhu replied:
Not sure I understand - isn't that fine (modifications you mentioned above notwithstanding)?

mhvk replied:
Just thought that for Mark4 at least, the sample rate should be in MHz.

mhvk commented on baseband/dada/base.py:
```diff
@@ -190,8 +191,7 @@ class DADAStreamBase(VLBIStreamBase):
     def __init__(self, fh_raw, header0, thread_ids=None, squeeze=True):
-        frames_per_second = (1. / (header0['TSAMP'] * 1e-6) /
-                             header0.samples_per_frame)
+        sample_rate = (1. / (header0['TSAMP'] * 1e-6)) * u.Hz
```
Since most bandwidths are more logically measured in `MHz`, let's do 
```
sample_rate = (1. / header0['TSAMP']) * u.MHz
```

mhvk commented on baseband/gsb/base.py:
```diff
@@ -223,24 +219,16 @@ class GSBStreamReader(GSBStreamBase, VLBIStreamReaderBase):
+                 bps=None, complex_data=None, samples_per_frame=None,
+                 payloadsize=None, squeeze=True):
         header0 = fh_ts.read_timestamp()
-        if frames_per_second is None and sample_rate is None:
```
I'm reconsidering. Let's just leave `samples_per_frame` at the end (and organize the other keywords to be in order as similar as possible to what is done for the other formats), and calculate properly below (or just raise an error if `None` for now).

mhvk commented on baseband/gsb/base.py:
```diff
@@ -320,9 +308,10 @@ def read(self, count=None, fill_value=0., out=None):
+                framerate = self.sample_rate / self.samples_per_frame
+                assert np.isclose(self._frame_nr,
+                                  ((self._frame.header.time -
+                                    self.start_time) * framerate).decompose())
```
`to(u.one)`

mhvk commented on baseband/vdif/header.py:
```diff
@@ -496,23 +499,30 @@ class VDIFSampleRateHeader(VDIFBaseHeader):
+    def sample_rate(self):
+        """Number of complete samples per second."""
+        return u.Quantity(self['sampling_rate'] *
+                          (1000000 if self['sampling_unit'] else 1000) *
```
Can we keep my expression above, which sets the unit appropriately (rather than multiply by 1000000 or 1000)? That way the `repr` will more naturally look OK.

Also, looking at this again, I worry whether one should indeed divide by `nchan` (I know I implemented it as such in framerate but I am not sure this is properly tested anywhere... could you re-check the VDIF documentation?).

cczhu replied:
Both EDV 1 and 3 specs say `sampling_rate` is a sample rate in ksps or Msps (so we shouldn't be dividing by `bps`, and we don't).  EDV 1's documentation compares the range of possible sampling rates to the K5/VSSP data acquisition systems (from 0.04 to 64 MHz).  A [K5 specification doc I found](ftp://ivscc.gsfc.nasa.gov/pub/TOW/tow2011/notebook/Kokado.Sem1.pdf) quotes sample rate values per channel (so the total data rate is `sample_rate * nchan * bps`), which suggests `sampling_rate` **shouldn't** be divided by `nchan` since it's already the per-channel rate.

However, for EDV 3, the convention is, "Each VDIF thread produced contains a single base-band channel", so `nchan = 1`.  EDV 1 doesn't make this specification, but it isn't explicit in what `sampling_rate` should be for multiple channels.

I'm worried this is actually an ambiguity in EDV 1 (the original VDIF specification doesn't have a sample rate, so that's okay).  Might pay to e-mail the NRAO folks.  In the meantime, this seems like the wrong PR to fix this issue (especially since it might break some other people's codes)?

Regardless, I've switched back to your expression above.

cczhu replied:
Actually, on second thought it's the right PR.  I'd still like confirmation from an external source, though.  Let's discuss it tomorrow at one of the meetings.

mhvk replied:
I am nearly sure we've encountered standard VDIF with single-thread, multiple channels and EDV=1, so I think the current implementation cannot be too wrong... Anyway, let's discuss outside of this PR -- goal now is simply to keep things consistent.

mhvk commented on baseband/vdif/header.py:
```diff
@@ -496,23 +499,30 @@ class VDIFSampleRateHeader(VDIFBaseHeader):
+    def sample_rate(self, sample_rate):
+        sample_rate = sample_rate.to(u.Hz)
+        assert sample_rate.value % 1 == 0
+        bandwidth = (sample_rate * self.nchan /
```
Again, just to be sure, please check the actual documentation...

mhvk commented on baseband/vdif/header.py:
```diff
@@ -671,10 +665,11 @@ def get_time(self, framerate=None, frame_nr=None):
+                                     "frame number requires a sample rate. "
                                      "Pass it in explicitly.")
+                framerate = sample_rate / self.samples_per_frame
                 offset = (frame_nr / framerate).to(u.s).value
```
I think I'd do (here and elsewhere) `offset = (frame_nr * self.samples_per_frame / sample_rate).to(u.s).value`, but it doesn't really matter.

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -496,13 +497,14 @@ def test_filestreamer(self):
             assert fh.tell() == 12
             t12 = fh.time
-            s12 = 12 / fh.samples_per_frame / fh.frames_per_second * u.s
+            s12 = (12 / fh.sample_rate).to(u.s)
```
No need to do `.to(u.s)`.

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -522,10 +524,10 @@ def test_filestreamer(self):
-            assert abs(fh.stop_time - fh.start_time - u.s * fh.size /
-                       fh.samples_per_frame / fh.frames_per_second) < 1. * u.ns
+            assert abs(fh.stop_time - fh._last_header.time - (
+                fh.samples_per_frame / fh.sample_rate).to(u.s)) < 1. * u.ns
```
No need for `.to(u.s)` (also two lines below)

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -648,33 +651,34 @@ def test_mwa_vdif():
 
 def test_arochime_vdif():
     """Test ARO CHIME format (uses EDV=0)"""
+    sample_rate = 5**8*u.Hz   # 390625 frames/second * 1 samples/frame.
```
If you want to understand where this comes from,
```
sample_rate = 800*u.MHz / 1024
```
(where one needs 1024=nchan*2 real samples for each FT)

mhvk commented on baseband/vdif/tests/test_vdif.py:
```diff
@@ -686,8 +690,7 @@ def test_arochime_vdif():
         assert abs(t1 - fh.stop_time) < 1. * u.ns
-        assert abs(t1 - t0 - u.s * (fh.size / fh.samples_per_frame /
-                                    fh.frames_per_second)) < 1. * u.ns
+        assert abs(t1 - t0 - (fh.size / fh.sample_rate).to(u.s)) < 1. * u.ns
```
No need for `.to(u.s)`

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -49,27 +48,19 @@ class VLBIStreamBase(VLBIFileBase):
 
     _frame_class = None
     _squeezed_shape = None
+    _samples_per_frame = None
```
I don't think we need to define these since you set them unconditionally in `__init__`. (Having them here suggests it might be relevant for subclasses to override them, but that is not the case).

cczhu commented on baseband/mark4/base.py:
```diff
@@ -320,16 +318,16 @@ def _get_frame_rate(fh, header_template):
         Returns
         -------
-        fps : int
+        framerate : int
```
I'd vote for returning a `Quantity` (since we make it one to get `sample_rate` anyway).  I'll revise `vlbi_base` as well.

cczhu commented on baseband/tests/test_conversion.py:
```diff
@@ -311,9 +309,11 @@ class TestDADAToVDIF1(object):
     """
+
     def get_vdif_header(self, header):
+        sample_rate = header.bandwidth * (1 if header.complex_data else 2)
```
Ha ha, I just did define it to fix an issue you brought up above (which for some reason I can't comment on).

Annoyingly, there's no way to force the user to consistently set `bandwidth` and `sample_rate`.  Can I dump `bandwidth` completely?

mhvk replied:
I think we had already decided to do this, but thought it was best done separately -- if it interferes here, let's indeed just dump it now.

cczhu replied:
Okay, took the liberty of removing `bandwidth`.

cczhu commented:
Addressed all PR comments.  Those that don't have comments were implemented into the code without issue.  For the DADA header, I added a `sample_rate` property that sets `BW` and `TSAMP` in the backend.  There's no simple way of setting `bandwidth` alongside this (it's possible to pass both to `DADAHeader.fromvalues` and write to `BW` and `TSAMP` twice) so I got rid of `bandwidth`.

I'm still confused as to the exact definition of `sampling_rate` in VDIF EDV 1 and 3.  See the in-line comment above for further details.

mhvk commented on baseband/dada/header.py:
```diff
@@ -336,15 +337,16 @@ def sample_shape(self, sample_shape):
+    @sample_rate.setter
+    def sample_rate(self, sample_rate):
+        sample_rate = sample_rate.to_value(u.MHz)
+        self['TSAMP'] = 1. / sample_rate
```
I think this should be `abs(sample_rate)` since as written one can still pass in a negative one to indicate sideband. I think this should be noted in the docstring...

mhvk commented on baseband/gsb/base.py:
```diff
@@ -415,10 +402,9 @@ class GSBStreamWriter(GSBStreamBase, VLBIStreamWriterBase):
     """
 
-    def __init__(self, fh_ts, fh_raw, header=None, nchan=None, bps=None,
+    def __init__(self, fh_ts, fh_raw, sample_rate, nchan=None, bps=None,
```
`header` should stay up front (after `sample_rate`)

cczhu replied:
For Mark 5 and VDIF, `header` is slotted just before `squeeze`.  I'll move it back to its previous position in GSB, and make a note in the issue we raised about this.

mhvk replied:
OK, sounds good. There is some sense in having `header` (for the writer) right after all the things that cannot be inferred from it, and before the keywords that might define it.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -426,16 +432,20 @@ def test_rawfile_io(self, tmpdir):
-        with gsb.open(SAMPLE_RAWDUMP_HEADER, mode='rs', raw=SAMPLE_RAWDUMP,
-                      payloadsize=2**12, squeeze=False) as fh_r:
+        with gsb.open(SAMPLE_RAWDUMP_HEADER, 'rs', raw=SAMPLE_RAWDUMP,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
use `self.payloadsize`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -515,8 +527,9 @@ def test_raw_stream(self, tmpdir):
-        with gsb.open(filename_incompletehead, mode='rs', raw=SAMPLE_RAWDUMP,
-                      payloadsize=2**12, squeeze=False) as fh_r:
+        with gsb.open(filename_incompletehead, 'rs', raw=SAMPLE_RAWDUMP,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
use `self.payloadsize`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -470,25 +480,27 @@ def test_raw_stream(self, tmpdir):
                       samples_per_frame=2**13) as fh_r, \
                 open(str(tmpdir.join('test_time.timestamp')), 'w+b') as sh,\
                 open(str(tmpdir.join('test.dat')), 'w+b') as sp:
             fh_w = gsb.open(sh, 'ws', raw=sp, samples_per_frame=2**13,
```
I'd do `self.payloadsize * (8 // bps)` (here and below) but it is not a big deal.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -525,9 +538,9 @@ def test_raw_stream(self, tmpdir):
-                      payloadsize=2**12, squeeze=False,
-                      frames_per_second=3.9736) as fh_r:
+        with gsb.open(filename_incompletehead, 'rs', raw=SAMPLE_RAWDUMP,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
`self.payloadsize`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -536,25 +549,30 @@ def test_raw_stream(self, tmpdir):
-        with gsb.open(SAMPLE_PHASED_HEADER, mode='rs', raw=SAMPLE_PHASED,
-                      payloadsize=2**12, squeeze=False) as fh_r:
+        with gsb.open(SAMPLE_PHASED_HEADER, 'rs', raw=SAMPLE_PHASED,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
`self.payloadsize`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -536,25 +549,30 @@ def test_raw_stream(self, tmpdir):
             fraw = [[open(thread, 'rb') for thread in pol]
                     for pol in SAMPLE_PHASED]
             with open(SAMPLE_PHASED_HEADER, 'rt') as ft:
                 frame1 = gsb.GSBFrame.fromfile(ft, fraw, payloadsize=2**12,
```
idem

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -536,25 +549,30 @@ def test_raw_stream(self, tmpdir):
             with open(SAMPLE_PHASED_HEADER, 'rt') as ft:
                 ft.seek(frame1.header.seek_offset(9))
                 self.seek_phased_rawfiles(fraw, 9 * fh_r._payloadsize)
                 frame10 = gsb.GSBFrame.fromfile(ft, fraw, payloadsize=2**12,
```
idem

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -583,20 +601,22 @@ def test_phased_stream(self, tmpdir):
+                      squeeze=False) as fh_r:
             fraw = [[open(thread, 'rb') for thread in SAMPLE_PHASED[1]]]
             with open(SAMPLE_PHASED_HEADER, 'rt') as ft:
                 frame1 = gsb.GSBFrame.fromfile(ft, fraw, payloadsize=2**12,
```
idem

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -583,20 +601,22 @@ def test_phased_stream(self, tmpdir):
-        with gsb.open(SAMPLE_PHASED_HEADER, mode='rs', raw=SAMPLE_PHASED,
+        with gsb.open(SAMPLE_PHASED_HEADER, 'rs', raw=SAMPLE_PHASED,
+                      sample_rate=sample_rate,
                       samples_per_frame=2**3) as fh_r, \
```
Like above, here and below, maybe `samples_per_frame=self.payloadsize / nchan`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -638,8 +659,9 @@ def test_phased_stream(self, tmpdir):
-        with gsb.open(filename_incompletehead, mode='rs', raw=SAMPLE_PHASED,
-                      payloadsize=2**12, squeeze=False) as fh_r:
+        with gsb.open(filename_incompletehead, 'rs', raw=SAMPLE_PHASED,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
`self.payloadsize`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -648,9 +670,9 @@ def test_phased_stream(self, tmpdir):
-                      payloadsize=2**12, squeeze=False,
-                      frames_per_second=3.9736) as fh_r:
+        with gsb.open(filename_incompletehead, 'rs', raw=SAMPLE_PHASED,
+                      sample_rate=sample_rate, payloadsize=2**12,
```
idem

mhvk commented on baseband/vdif/base.py:
```diff
@@ -238,37 +238,36 @@ class VDIFStreamBase(VLBIStreamBase):
     def _get_time(self, header):
         """Calculate time for given header.
 
         This passes on frame rate, since not all VDIF headers can calculate it.
```
Update docstring

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -188,19 +204,20 @@ def tell(self, unit=None):
     def _frame_info(self):
         offset = (self.offset +
                   self.header0['frame_nr'] * self.samples_per_frame)
+        framerate = int(np.round((
```
Really trivial, but maybe move last parenthesis to next line, to make it clearer what is being rounded.
